opentracing/index.js

```diff
@@ -0,0 +1,188 @@
+'use strict'
+
+const opentracing = require('opentracing')
+const Noop = require('opentracing/lib/noop')
+const ElasticTraceContext = require('../lib/instrumentation/trace-context')
+const Agent = require('../lib/agent')
+const Transaction = require('../lib/instrumentation/transaction')
+
+const illigalChars = /[.*"]/g
+
+class Tracer extends opentracing.Tracer {
+  constructor (agent) {
+    super()
+    if (agent && agent instanceof Agent) {
+      this._agent = agent
+    } else {
+      this._agent = require('../').start(agent)
+    }
+  }
+
+  _startSpan (name, opts) {
+    let context = null
+
+    if (Array.isArray(opts.references)) {
+      if (opts.references.length > 1) {
+        this._agent.logger.debug('Elastic APM OpenTracing: Unsupported number of references to _startSpan:', opts.references.length)
+      }
+
+      for (const i in opts.references) {
+        const ref = opts.references[i]
+        const refType = ref.type()
+        // TODO: Log warning if type is REFERENCE_FOLLOWS_FROM?
+        if (refType === opentracing.REFERENCE_CHILD_OF) {
+          context = ref.referencedContext()._elasticContext
+          break
+        }
+      }
+    }
+
+    let tags, type
+
+    if (opts.tags) {
+      // We might delete tags leter on, so clone in order not to mutate
+      tags = Object.assign({}, opts.tags)
+      type = tags.type
+      if (type) delete tags.type
+    }
+
+    let span = this._agent.currentTransaction === null || this._agent.currentTransaction.ended
+      ? this._agent.startTransaction(name, type, { traceContext: context, startTime: opts.startTime })
+      : this._agent.startSpan(name, type, { traceContext: context, startTime: opts.startTime })
+
+    span = span ? new Span(this, span) : Noop.span // TODO: What happens when a user tries to use the context of a Noop span? Maybe the Noop span context should be the context of the transaction?
+
+    if (tags && Object.keys(tags).length !== 0) span._addTags(tags)
+
+    return span
+  }
+
+  _inject (spanContext, format, carrier) {
+    switch (format) {
+      case opentracing.FORMAT_TEXT_MAP:
+      case opentracing.FORMAT_HTTP_HEADERS:
+        carrier['elastic-apm-traceparent'] = spanContext.toString()
+        break
+    }
+  }
+
+  _extract (format, carrier) {
+    let ctx
+
+    switch (format) {
+      case opentracing.FORMAT_TEXT_MAP:
+      case opentracing.FORMAT_HTTP_HEADERS:
+        ctx = ElasticTraceContext.fromString(carrier['elastic-apm-traceparent'])
```

Comment:
throws when the elastic-apm-traceparent header is undefined. 

```js
Elastic APM caught unhandled exception: Cannot read property 'replace' of undefined
TypeError: Cannot read property 'replace' of undefined
   at headerToBuffer (/Users/vigneshh/opensource/tailor-apm-demo/node_modules/elastic-apm-node/lib/instrumentation/trace-context.js:83:23)
    at Function.fromString (/Users/vigneshh/opensource/tailor-apm-demo/node_modules/elastic-apm-node/lib/instrumentation/trace-context.js:136:29)
    at Tracer._extract (/Users/vigneshh/opensource/tailor-apm-demo/node_modules/elastic-apm-node/opentracing/index.js:75:35)
    at Tracer.extract (/Users/vigneshh/opensource/tailor-apm-demo/node_modules/opentracing/lib/tracer.js:122:21)
```

 The usage is done like this 

https://github.com/zalando/tailor/blob/52812a3ebaff911524d683377872e23389bb9a69/lib/request-handler.js#L38-L41

Reply:
Good catch 😅 

We just moved the opentracing bridge to it's own module: https://github.com/elastic/apm-agent-nodejs-opentracing

I'll fix it there 👍